server.go

```diff
@@ -261,6 +271,160 @@ HANDLE:
 	goto LOOP
 }
 
+func (s *Server) handleUnlimitedPacket(frisbeeConn *Async, connCtx context.Context) {
+	p, err := frisbeeConn.ReadPacket()
+	if err != nil {
+		_ = frisbeeConn.Close()
+		s.onClosed(frisbeeConn, err)
+		return
+	}
+	if s.ConnContext != nil {
+		connCtx = s.ConnContext(connCtx, frisbeeConn)
+	}
+	wg := new(sync.WaitGroup)
+	closed := atomic.NewBool(false)
+	connCtx, cancel := context.WithCancel(connCtx)
+	handler := func(p *packet.Packet) {
+		handlerFunc := s.handlerTable[p.Metadata.Operation]
```

Comment:
I'd use `defer wg.Done()` here

Reply:
There's an unfortunate overhead to using `defer` - it's not that bad but if we can avoid it I'd prefer to:

- https://github.com/hashicorp/golang-lru/pull/53
- https://github.com/golang/go/issues/14939

Reply:
Unless there really is a big penalty I'd say it makes code more readable & robust against 'failed to cleanup on all function returns' bugs.

Out of date issue though afaik.

In golang release 1.14: "This release improves the performance of most uses of defer to incur almost zero overhead compared to calling the deferred function directly. As a result, defer can now be used in performance-critical code without overhead concerns. "

Reply:
Let me make those changes and run some benchmarks

Reply:
Okay so I went into our code base for `async.go` (which is where the majority of `defer` isn't being used) and it looks like it's done that way on purpose. I need to actually call wg.Done before I call close sometimes (since `close` calls `wg.Wait`). 

This means adding `defer` statements requires much more thought than I first realized.

While I totally agree that not having the defers would be better for now we are checking for leaking goroutines in our test cases in `helpers_test.go/TestMain()` - if we miss a wg.Done or don't clean up one of our return functions properly the goleak checker should catch that. 

All that being said, I think it would be beneficial for us to create an issue to investigate this further.

The issue has been created here: https://app.zenhub.com/workspaces/loophole-labs-61d69c9aa0a1140011fbdff4/issues/loopholelabs/frisbee-go/147